packages/medusa/src/services/order.js

```diff
@@ -0,0 +1,439 @@
+
+    /** @private @const {FulfillmentProviderService} */
+    this.fulfillmentProviderService_ = fulfillmentProviderService
+
+    /** @private @const {EventBus} */
+    this.eventBus_ = eventBusService
+  }
+
+  /**
+   * Used to validate order ids. Throws an error if the cast fails
+   * @param {string} rawId - the raw order id to validate.
+   * @return {string} the validated id
+   */
+  validateId_(rawId) {
+    const schema = Validator.objectId()
+    const { value, error } = schema.validate(rawId)
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "The order id could not be casted to an ObjectId"
+      )
+    }
+
+    return value
+  }
+
+  /**
+   * Used to validate order addresses. Can be used to both
+   * validate shipping and billing address.
+   * @param {Address} address - the address to validate
+   * @return {Address} the validated address
+   */
+  validateAddress_(address) {
+    const { value, error } = Validator.address().validate(address)
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "The address is not valid"
+      )
+    }
+
+    return value
+  }
+
+  /**
+   * Used to validate email.
+   * @param {string} email - the email to vaildate
+   * @return {string} the validate email
+   */
+  validateEmail_(email) {
+    const schema = Validator.string().email()
+    const { value, error } = schema.validate(email)
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "The email is not valid"
+      )
+    }
+
+    return value
+  }
+
+  /**
+   * Gets an order by id.
+   * @param {string} orderId - id of order to retrieve
+   * @return {Promise<Order>} the order document
+   */
+  async retrieve(orderId) {
+    const validatedId = this.validateId_(orderId)
+    const order = await this.orderModel_
+      .findOne({ _id: validatedId })
+      .catch(err => {
+        throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+      })
+
+    if (!order) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Order with ${orderId} was not found`
+      )
+    }
+    return order
+  }
+
+  /**
+   * Creates an order
+   * @param {object} order - the order to create
+   * @return {Promise} resolves to the creation result.
+   */
+  async create(order) {
+    return this.orderModel_.create(order).catch(err => {
+      throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+    })
+  }
+
+  /**
+   * Updates an order. Metadata updates should
+   * use dedicated method, e.g. `setMetadata` etc. The function
+   * will throw errors if metadata updates are attempted.
+   * @param {string} orderId - the id of the order. Must be a string that
+   *   can be casted to an ObjectId
+   * @param {object} update - an object with the update values.
+   * @return {Promise} resolves to the update result.
+   */
+  async update(orderId, update) {
```

Comment:
Good points!

This is actually why I think it would be reasonable to keep `updateShipping` and `updateBilling`... Otherwise, the update function would need to check for quite a lot of things.

We still only keep ONE `update-order` endpoint, and then we just use the dedicated methods in that endpoint - like carts for example. 

What are your thoughts on this?

Reply:
EDIT: I feel like my last comment was not saying much.

If we need to make multiple combinatory checks in the update. function, I would find it alright to keep dedicated methods for those specific fields to update. For example shipping and billing. Those would both need to actually validate the data. But then also check for the different combinations of order status. 

It simply only to avoid cramping up the update method, but I am really indifferent on the matter

Reply:
Actually, you know what. Let's only have one `update` method haha. It makes sense. :D Sorry for the rambling

Reply:
> Specifically we need to ensure that the order has not been shipped or captured otherwise it doesn't make sense to update certain fields. For example, we shouldn't allow shipping address updates if the order has been shipped.

The update operation constraints for billing and shipping are the same right? Cause when an order is fulfilled or captured, you can't change either.

Reply:
Haha glad I made you think this through :) 

An easy way to make a separation of concerns within the service would be to make the `update[Field]` methods private. That way you could still have nice methods that can be unit tested easily, while your `update` would be more concise with something like:

```
if (!_.isEmpty(update.shipping_address)) {
  this.updateShippingAddress_(orderId, update.shipping_address)
}
```

It would however be nice to have the database operation be one single `updateOne` call so maybe something like: 
```
let updateFields = {}
if (!_.isEmpty(update.shipping_address)) {
  updateFields.shipping_address = this.updateShippingAddress_(orderId, update.shipping_address) // returns validated shipping address update
}
this.model_.updateOne(orderId, { $set: updateFields })
```

Reply:
Yeah, I like this idea!